Update README.md and split into several related files

ThanGerlek opened the pull request:
A PR for #451. Adds, updates, clarifies, and reorganizes "getting started" info for new devs joining the autograder team.

- Extract getting started info from `README.md`
- Reorganize information by use case (e.g. isolate Windows-specific info)
- Remove/replace outdated information
- Add relevant "getting started" info
- General improvements to setup instructions

ThanGerlek commented:
I removed as much incorrect/outdated info as I could, but there were several things I wasn't sure about, and I'm sure I missed some. Keep an eye out for anything I mistakenly changed

frozenfrank commented:
Thanks @ThanGerlek! I'll review this when I have a moment.

Quickly though, you may noticed that I added a new line to the end of your PR description. It contains a special [closing keyword](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue) which GitHub recognizes to directly link this PR to the issue you opened earlier (great job BTW). These closing keywords only work when placed inside the PR description (putting it in a comment won't do the trick).

You'll also notice that I added a link to the issue in #444 which is the parent issue of these kinds of changes. Previously, this kind of adjustment had been reference by text only, but I hadn't yet created an issue for it. Thank you for doing so!

frozenfrank commented:
Upon closer examination of the sources, I understood why you had not wanted this to **close** the #451 issue: You intend for future work to actually clean up the content whereas this PR mainly affects the organization of the files.

I removed the closing keywords from the PR description accordingly.

frozenfrank commented on getting-started/getting-started.md:
```diff
@@ -0,0 +1,135 @@
+
+Operating System-specific instructions (read the applicable doc before continuing):
+
+- [Windows](windows.md)
```
There is currently no need for special instruction pages for Linux and MacOS devices. According to the [YAGNI](https://martinfowler.com/bliki/Yagni.html), we should not include those pages. We don't even expect to need them!

We could replace this section with the following Windows specific message (which displays as an appropriate warning box for users).

```md
> [!WARNING]
> The autograder unfortunately requires special setup instructions
> in order to operate on **Windows machines**.
> 
> Read [Getting Started on Windows](windows.md) before continuing.
```

ThanGerlek replied:
Agreed, but I left them in case @19mdavenport or any of the other TAs have useful information to add. I only know about Windows 🤷‍♂️

frozenfrank commented on README.md:
```diff
@@ -46,262 +46,20 @@ property `db.url` is set to `db:3306`
+- Review the Sequence/Class Diagrams (_coming soon_) and try creating one
+- Add documentation
+- Find a section of under-tested code and add some unit tests
+- Take a look through the GitHub repo's Issues page and find one you like, especially ones labeled "good first issue"
```
We can provide a link to a filtered view of the issues page which only displays those with the referenced tag.

Let's add this link over the words "good first issue": 
https://github.com/softwareconstruction240/autograder/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22

frozenfrank commented on getting-started/windows.md:
```diff
@@ -0,0 +1,60 @@
+
+###### VSCode
+
+Go fish🐟 These instructions are not included in this file. It should be possible to simply use VSCode's Dev Containers
```
See earlier comment about removing recommendations to setup with VSCode.

frozenfrank reviewed (changes requested):
This structural change looks great, @ThanGerlek, and I really appreciate the expanded and clarified content along the way!

I've provided several comments in files and lines of code with ways we may be able to improve the presentation of information.

After working through all the content, I realized that I ended up proposing two opposing ways of dealing with the platform-specific setup instructions. Both sets of comments remain in the review, but I'll clarify which I recommend. The opposing paths are:
1. Remove unused/unneeded setup instruction files (recommended)
2. Change content of filler files (**not** recommended)

frozenfrank commented on README.md:
```diff
@@ -46,262 +46,20 @@ property `db.url` is set to `db:3306`
+First read the relevant OS-specific instructions:
 
-##### Dev Container (recommended for Windows)
+- [Windows](getting-started/windows.md)
```
Since this information is already contained in the `getting-started/getting-started.md` file, and because the information is more appropriately contained in that file, I recommend removing it entirely from this root `README.md` file.

Subsequent changes to the wording of the next sentence would also be appropriate.

ThanGerlek commented on getting-started/getting-started.md:
````diff
@@ -0,0 +1,135 @@
+```
+
+While you can use any root user credentials to access the MySQL database, you may be interested in creating
+a special login for this project with restricted privileges (DELETE and CREATE USER administrator privileges are
````
I don't see this... What proposed change are you referring to?

frozenfrank replied:
I added the word "USER" after "DELETE" to be parallel to "CREATE USER."

I don't actually know which technical entities are being referred to in the statement though, and I could be wrong.

I searched online for SQL statements relating to CREATE/DELETE user, but I found `DROP LOGIN` and `DROP USER` instead. Can we get a link to the "administrator privileges documentation that describes which restricted privileges are required?

ThanGerlek replied:
Right. I'm blind.

"CREATE USER" and "DELETE" are just the exact names of the checkboxes I needed to click on in MySQL Workbench (under Users and Privileges/Administrative Roles/Global Privileges) in order to make it work. I have no idea what MySQL codes they map to. But that was last year, and some experimentation today revealed you need pretty much everything anyway (it needs to be able to call GRANT ALL on the student user, so it needs at least "GRANT ALL"-level privileges). So listing them here really isn't useful anymore.

ThanGerlek commented on README.md:
```diff
@@ -46,262 +46,20 @@ property `db.url` is set to `db:3306`
-click a few buttons along the way. Eventually a new IntelliJ window will pop up from the dev container. Follow the
-[Yarn](#nodeyarn) instructions and [Getting Started](#getting-started) steps below with the new window. 
-Use `host.docker.internal` as your db-host argument.
+Then follow the [Getting Started](getting-started/getting-started.md) page to Get Started!
```
Eh, I was tryna be clever. It's changed.

ThanGerlek commented on getting-started/windows.md:
```diff
@@ -0,0 +1,60 @@
+
+The autograder unfortunately won't work directly from Windows, so it must be run inside a Docker container
+(recommended) or inside WSL (possible but can be finicky). When running the autograder inside a Docker container,
+you will need to use a _Dev Container_ if you want to use IntelliJ's or VSCode's debugging tools.
```
I've added some introductory info. Do you think it's clear?

frozenfrank commented on getting-started/db-insert-statements/insert-config-database.md:
Those are important things to consider. Given that additional context, recognizing that new users are only expected to use one of them, I think it's fine to leave "as is" until we reorganize the insert statement organizations.

frozenfrank commented on getting-started/windows.md:
```diff
@@ -1,22 +1,26 @@
 The autograder unfortunately won't work directly from Windows, so it must be run inside a Docker container
-(recommended) or inside WSL (possible but can be finicky). When running the autograder inside a Docker container,
-you will need to use a _Dev Container_ if you want to use IntelliJ's or VSCode's debugging tools.
+(recommended) or inside WSL, the Windows Subsystem for Linux (possible but can be finicky). When running the autograder
```
Thank you for these clarifications!

frozenfrank commented on getting-started/windows.md:
```diff
@@ -36,17 +40,24 @@ and paste a link to the autograder GitHub repo (or your fork of it). The remaind
-that everything in these instructions should be done via a WSL terminal _and_ inside a WSL directory. If you use a
-Windows directory, like Desktop/chess, you might run into file permissions errors. "Unix-like" shells, such as Git
-Bash, will not work: it needs to be a true WSL terminal.
+*WARNING*: Everything in these instructions (including everything in `getting-started.md`)
```
I'll convert these to a special GitHub alert callout.

frozenfrank reviewed (commented):
These changes are looking great!

There's a presentation change I want to make; but otherwise, a big thank you for making these changes happen!

frozenfrank reviewed (approved):
This looks good to me, but I would wait for a review by @19mdavenport before merging it.

What else would be needed in order to actually close #451? (Have we made the adjustments already; if not, can you specifically update #451 to indicate what work remains?)

frozenfrank commented:
@ThanGerlek FYI It seems that we keep fighting over trailing whitespace on the lines.

Can you make sure you have these "On Save" settings enabled?
* Remove trailing whitespace
* Remove trailing blank lines at the end of saved files
* Ensure every saved file ends with a line break

<img width="975" alt="image" src="https://github.com/user-attachments/assets/9d01b72c-65b5-4782-a348-63e0b4f49e8c">

ThanGerlek commented:
Done. Thanks for catching that. Maybe a code style XML or something would be useful for the autograder as well as the chess project... not exactly top priority though

frozenfrank reviewed (commented):
OK, now I'm fully comfortable with the presentation of the content. I feel like the information is a readable as it can be, and I'm happy with the way it presents on the screen.

I did chat with @19mdavenport this morning, and he mentioned that there may be some content adjustments we need to make. Let's wait to hear back from him— he's a busy man, but he'll look at this soon enough.

webecke commented:
Just a small thing, GitHub treats the `docs/` folder specially, especially in a way that will help with work going on in #448. Should we put all these files in a `docs/` folder or do we specifically want it to be in a separate `getting-started/` folder?

frozenfrank commented:
> Just a small thing, GitHub treats the `docs/` folder specially, especially in a way that will help with work going on in #448. Should we put all these files in a `docs/` folder or do we specifically want it to be in a separate `getting-started/` folder?

Great suggestion, @webecke! This is accounted for in 571cca5de05fea11072bc78b8ccf9ee8c1cba618.

frozenfrank commented:
Also marking this as resolving #335 since the clarifications and adjustments to the language indicate no further requirement to have Canvas API tokens.

webecke commented:
I consolidated the development and getting started sections of the README.md into one small section and moved it to the top. If we're opposed to that we can change it back, but for almost everyone coming to look at the Autograder repo the info they need is in the Getting Started Guide and Contribution Guide. Thus I propose they should be at the top.

ThanGerlek commented:
Updated the PR description to represent recent changes.

At this point this PR will probably close issue #452, although not until we're all happy with it, so I've changed it to a draft PR as well.

frozenfrank commented:
@19mdavenport I think you're review is the last thing we're looking for on this PR.

@ThanGerlek Are you aware of any other changes we're still waiting for?

ThanGerlek commented:
Nope, that's all. Not just Michael's review but also any additions he thinks might be useful

frozenfrank commented:
> Nope, that's all. Not just Michael's review but also any additions he thinks might be useful

I went ahead and marked it as "ready to review" since you have contributed everything you are planning on for this PR. Other reviewers and contributors may weigh in, but now is the time for that process to be happening. (Typically, the "draft" stage is a signal that other people should _not_ look at this PR because the PR author is still working on critical aspects of it.)

webecke commented:
I vote we go ahead and merge this in if no one else has things to add. @frozenfrank @ThanGerlek @19mdavenport
